XXX-module-import.md

```diff
@@ -0,0 +1,241 @@
+| Title  | module.import(path):Promise |
+|--------|-----------------------------|
+| Author | @WebReflection              |
+| Status | DRAFT                       |
+| Date   | 2017-01-23                  |
+
+# Asynchronous `module.import(path)`
+There is [a standard effort](https://github.com/tc39/proposal-dynamic-import)
+to bring asynchronous module loading to JavaScript.
+
+Such effort will take some time before it will be fully agreed,
+and before it will be widely available across all JS engines.
+
+In the meanwhile, the community could benefit from this _opt-in_
+proposal, which is 100% backward compatible, as well as future friendly,
+being based entirely on what's been agreed already by TC39.
+
+## The pseudo implementation example
+The following code, which works already,
+represents how essential is this additional
+`module.import` method proposal:
+```js
+// one added method to Module.prototype
+module.constructor.prototype.import =
+  function (path) {
+    return Promise.resolve()
+           .then(() => this.require(path));
+  };
+```
+
+### Fully backward compatible
+Every existent NodeJS/CommonJS module can be imported already.
+There is no need to change anything to the existing code-base.
+```js
+// even core modules work already
+module.import('fs').then(fs => {
+  // do something with fs module
+  fs.readdir('.', console.log);
+});
+```
+
+### Multiple imports too
+Keeping the implementation as simple as possible in order
+to have zero conflicts with current dynamic
+standard `import` proposal, we can use standard
+`Promise.all` method to retrieve at once every needed module.
+```js
+Promise.all([
+  module.import('fs'),
+  module.import('path')
+]).then(([fs, path]) => {
+  fs.readdir(
+    path.join(__dirname, 'test.js'),
+    console.log
+  );
+});
+```
+
+It's eventually possible to shortcut multiple imports
+using the following pattern.
+However, it won't be fully compatible
+with current standard `import` proposal as it is today.
+```js
+Promise.all([
+  'fs',
+  'path'
+].map(
+  m => module.import(m)
+)).then(modules => {});
+```
+
+## Unleashing asynchronous exports
+While good old modules will keep being usable,
+new modules might define themselves as asynchronous
```

Comment:
> but no one will do that. 

It's like saying: this only works if everyone will use `import()` instead of `require()` but no one will do that.

I don't like argumentations based on speculations. This is a moving-forward pattern, someone might do it, specially on the web.

One of the clear goal is improving universal modules.

Synchronous `require` is why we need bundlers, AMD was born before Promise were there and now we have a standard proposal based on Promise.

I won't comment further about speculations on what you think everyone does or want.

I hope this can be kept as regular conversation and exchange of ideas.
Thank you